introduce happens-before notion into the ledger model (#21563)
digital-asset/daml: sdk/docs/manually-written/overview/explanations/ledger-model/ledger-structure.rst

Reviewer: Not directly related to the changes in this PR: "For another example, consequences of an Exercise action form a transaction. In the example of the Settle action on Alice’s and Bob’s SimpleDvP, the consequences of the Settle action form the following transaction [..]" Isn't the transaction formed by the `Settl`e action, and then the two exercise subactions, i.e., the consequences of `Settle`, form a subtransaction?

Reviewer: Nit: It'd be nice if 1.4 reiterated why defining subtransactions is relevant, i.e., because our notion of privacy is at the level of subtransactions.

Author: > Isn't the transaction formed by the `Settle` action, and then the two exercise subactions, i.e., the consequences of `Settle`, form a subtransaction?

Formally, a transaction is a list of actions. As such, the list with the `Settle` action as its single element is a transaction. The consequences of `Settle` are also a list of actions and therefore by definition a transaction.

It is correct that the consequences of `Settle` are a subtransaction, but that does not diminish the fact that they are formally a transaction. In particular, the subtransaction relationship is introduced only later.

Reviewer: Perhaps the disconnect is that I see a transaction as an indivisible unit. Even if definition-wise it might be correct, it might raise some questions what it means guarantee-wise that a subtransaction - which should not exist standalone (although a participant can in fact commit only a leg of a transaction in some corner cases, right) is defined as a transaction.

Ledger
******

> The transaction structure records the contents of the changes.

Author: `Changes` is used throughout the ledger model. It's already brought up in the first paragraph of this page and on the `index.rst` page as the concept of what the ledger records. I don't want to specialize it to a technical term like `actions` because that would miss the point, namely referring back to the informal concept of what is recorded. If you prefer, `ledger changes` would be more specific here.

> The transaction structure records the contents of the changes.

Reviewer: Sorry, I don't follow. `Change` is ok in the first paragraph because we haven't defined much yet. However, after the section goes to great lengths to define actions and transactions, I personally find it unclear why we need to resort to the generic term of `change` in the ledger section. I would have expected that changes to the ledger can be expressed very concretely in terms of transactions, actions. Put differently, do you think the reader won't wonder: "what change at the ledger level does transaction x create"?

> The transaction structure records the contents of the changes.

Author: We settled on "party interaction" instead of "change".

> The ledger data records two more aspects of a change:

Reviewer: ```suggestion
The ledger data records two more aspects of an action:
```

Author: Those aspects are not part of an action, but of a change that the ledger records.

* An identifier to uniquely refer a particular change.

> * The parties who requested a particular change.

Author: Same answer.

> Due to the :ref:`privacy model <da-model-privacy>`, not everyone sees all parts of a change.

Author: A change is neither an action nor a transaction. It's the informal concept of what needs to be recorded on the ledger.

> A unique identifier for a change allows different parties to correlate the changes they see.

Author: No. In general parties will see different projection of a transaction. These projections are transactions themselves and that's what parties see. In particular, the parties do not see the original transaction. Yet, they can correlate via the update ID whether their projections belong to the same ledger change, i.e., are subtransactions of the original commit.

> A unique identifier for a change allows different parties to correlate the changes they see.

Reviewer: My point was different: "correlating changes" can be interpreted by the reader as a participant can tell which change happens first. Here the participant can only tell that a (projection of a) txn originates from the same txn or not.

> The notion of an **update** adds such an identifier.

Reviewer: At what level does the `update` notion exist, is it at the ledger level? Until now, the section discussed concepts that emerge from exercising choices and whether choices are exercised at the same time.

Author: Updates exist at the ledger level. That's why this paragraph is in the `Ledger` section. Do you think there's a sign post missing?

It consists of a single transaction and the so-called **update ID**, a string.
Examples in the ledger model use update IDs of the form ``TX i`` like the transaction view in Daml Studio,

> whereas the update IDs on the Ledger API are hex strings.

Author: > Moreover: Does Daml Studio invent its own txn identifiers different from the LAPI ones, or is `i` in `Tx i` a hex string?

Daml Studio uses `TX 0`, `TX 1`, etc., which is different from what you get on the LAPI.

> whereas the update IDs on the Ledger API are hex strings.

Reviewer: I see, then it makes sense to keep the sentence: Depending on what the reader used or interacted with (LAPI, Daml studio), they'll see different tx ids, and this section just mentions that. Perhaps rephrasing as "Readers using Daml Studio will see, whereas readers using LAPI will see" or something like that.

A **commit** adds the information *who requested a change*.
It consists of an update and the one or more parties that requested it.
Those parties called the **requesters** of the commit.
In Daml Script, the requesters correspond to the ``actAs`` parties given to the ``submit`` commands.

> Definition **Ledger**:

Author: I do want a box! Thanks for bringing this up again. I finally found a suitable directive: `admonition`!

>   A **Ledger** is a finite directed acyclic graph (DAG) of commits, where the update IDs are unique.

Reviewer: I agree, but I don't see how it answers my question. Why do we need to specify "the ledger is a *finite* directed acyclic graph (DAG) of commits", as opposed to simply skipping finite: "the ledger is a directed acyclic graph (DAG) of commits"

>   A **Ledger** is a finite directed acyclic graph (DAG) of commits, where the update IDs are unique.

Author: >  I don't see how it answers my question

Then I don't understand your original question. Let's discuss offline.

> Definition **top-level action**:

Reviewer: Same as https://github.com/digital-asset/daml/pull/21563/files#r2222608749

>    :alt: The sequence of commits for the whole DvP workflow. First, banks 1 and 2 issue the assets, then Alice proposes the DvP, and finally Bob accepts and settles it.

Author: I see four commits in the diagram with update IDs `TX 0` through `TX 3`. What is missing in your view?

>    :alt: The sequence of commits for the whole DvP workflow. First, banks 1 and 2 issue the assets, then Alice proposes the DvP, and finally Bob accepts and settles it.

Author: Fixed the literal include.

.. note::
   The integrity constraints do not impose an order between independent commits.
   In this example, there need not be edges among the first three commits ``TX 0``, ``TX 1``, and ``TX 2``,
   so they could be presented in any order.

>    As the Ledger is a DAG, one can always extend the order into a linear sequence via a topological sort.

Reviewer: ```suggestion
   As the Ledger is a DAG, one can always extend the order into a linear sequence via a topological sort over the transaction identifiers.
```
